src/libmeasurement_kit/nettests/runnable.cpp

```diff
@@ -238,74 +238,131 @@ std::string Runnable::generate_output_filepath() {
     return filename.str();
 }
 
+void Runnable::contact_bouncer(Callback<Error> cb) {
+    if (!needs_bouncer) {
+        cb(NoError());
+        return;
+    }
+    auto bouncer = options.get(
+        "bouncer_base_url", std::string{"https://a.collector.ooni.io/bouncer"});
+    logger->info("Contacting bouncer: %s", bouncer.c_str());
+    ooni::bouncer::post_net_tests(
+        bouncer, test_name, test_version, required_test_helpers,
+        [=](Error error, Var<BouncerReply> reply) {
+            if (error) {
+                cb(error);
+                return;
+            }
+            assert(!!reply);
+            auto maybe_collector = reply->get_collector_alternate("https");
+            if (!maybe_collector) {
+                cb(maybe_collector.as_error());
+                return;
+            }
+            logger->info("Bouncer discovered collector: %s",
+                         maybe_collector->c_str());
+            if (options.find("collector_base_url") == options.end()) {
+                options["collector_base_url"] = *maybe_collector;
+                logger->info("Using discovered collector");
+            }
+            for (auto th: required_test_helpers) {
+                auto maybe_backend = reply->get_test_helper_alternate(
+                    th, "https"
+                );
+                if (!!maybe_backend) {
+                    logger->info("Bouncer discovered backend: %s",
+                         maybe_backend->c_str());
+                    if (options.find("backend") == options.end()) {
+                        logger->info("Using discovered backend");
+                        options["backend"] = *maybe_backend;
+                        // FIXME: This breaks multi-backend tests!!!
```

Comment:
Maybe a possible (not requested :) solution would be to make **std::list<std::string> required_test_helpers** a map instead of a list and then fill the values if we find them.